### PR TITLE
[Modelling] ROPE and Prompt Cross-Attention

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
ROPE:
* Applied to the q/k/v states in the self-attention
* Applied to the q states only in the cross-attention (not the k/v states)
* The rationale is that the k/v states come from the encoder, which has T5 positional embeddings already applied

Cross-Attention:
* Option to concatenate the T5 encoder hidden-states and prompt embeddings to be used as cross-attention conditioning
* If we do this, we no longer have to concatenate the prompt embeddings to the input embeddings
* We also apply a positional embedding to the prompt embeddings to encode positional info

### CODE REVIEW

File: parler_tts/modeling_parler_tts.py [hunk 1]
```diff
@@ -867,6 +958,36 @@ def forward(
...
+            # embed positions
+            # TODO: As it is, the masked ids from the prompt will still count in the positions embeddings
+            # maybe should modify position embeddings
+            positions = self.embed_positions(inputs_embeds, past_key_values_length)
+            hidden_states = inputs_embeds + positions.to(inputs_embeds.device)
+        else:
+            hidden_states = inputs_embeds
+
+            if position_ids is None:
+                if attention_mask is not None:
```
Comment (sanchit-gandhi) [inline parler_tts/modeling_parler_tts.py:972]: Ideally this logic would go in `prepare_inputs_for_generation` (e.g. if we did a formal Transformers integration). The reason we don't is that we do some pre-processing of the `prompt_attention_mask` and `attention_mask`, from which we compute the `position_ids`. This pre-processing currently only happens in the `.forward` pass, so it's cleaner to also compute the `position_ids` here as well

**Comment by ylacombe** [comment]
I think I've messed up merging, will fix later, sorry for that

File: parler_tts/modeling_parler_tts.py (hunk 1 above)
Comment (ylacombe) [reply parler_tts/modeling_parler_tts.py:1358]: So we create `position_ids` at every generation step ?

**Comment by ylacombe** [comment]
#65 superseded this, thanks for the effort @sanchit-gandhi !

**Pull request closed without merging**